Recipe class2.0

Sophiak113 opened the pull request:
Changed recipe class according to comments. Still needs tweaking

Sophiak113 committed:
Created Recipe class with attributes

Sophiak113 committed:
Added description getter and setter and tested it

Works so far

Sophiak113 committed:
Added input validation and another getter and setter

Added input validation to make sure an integer has to be entered. And tested it, it works. Added another getter and setter for number of servings.

Sophiak113 committed:
Changed naming convention

From camel case to lowercase with underscore

Sophiak113 committed:
Added more getters and setters and tested

Added getter and setter for calories per person and get ingredients. Tested to check if ingredients works

Sophiak113 committed:
Added instructions getter and setter and questioned data type

I put instructions as a list data type to possibly list step by step but was conflicted whether or not to go for string

Sophiak113 committed:
Added methods to remove ingredients and instructions

Sophiak113 committed:
Imported tkinter and checked if methods worked with it

Also experimented with whether tkinter needs text= to display

Sophiak113 committed:
Denoted what is test code

Therefore will not be in the final program

Sophiak113 committed:
Discovered display method would not work with tkinter

get method works with tkinter may not need display method

Sophiak113 committed:
Made attributes private and moved test code to separate file

Sophiak113 committed:
Set input type

Sophiak113 committed:
Deleted Test code and put duplicate prevention on set ingredients

Sophiak113 committed:
Trouble with display ingredients, added notes and minor tweaks

Changed line 38 to remove ingredient function. Put notes where code needs work

lachlanin4 commented on recipe.py:
```diff
@@ -34,9 +34,13 @@ def get_ingredients(self):
         self._ingredients.append(ingredients)
-
+        if len(set(self._ingredients)) != len(self._ingredients):
+            self._ingredients.remove(ingredients) #I have tested this and it only removes the recently added ingredient
```
If its a straightforward set then it should be a simple assignment I think - will probably update when the ingredients class is added in.

Sophiak113 replied:
I am getting confused between if you mean set the data type and set the function? Due to ingredients class does this mean only the self._ingredients.append(ingredients) is needed for this function?

lachlanin4 replied:
So the function is "set_ingredients" a typical interpretation could be that this sets either a subset or all of the ingredients. If you set a subset then this is either modifying a subset of ingredients that are already present or adding a subset (with an expectation of adding more) - or in the other case you are setting all of them. If you want to modify a subset of items in a list say what is typically done is you get the current list (as a copy) modify the items you need to (keeping everything else the same) and then overwrite the original list with the modified one you create. If the list is long and complicated then yes this can become cumbersome and a more sophisticated method of *in place" replacement is required. In this case I don't think the case really requires it. If everything needs to be set then one way or another the original list will be replaced completely so piecemeal partial modification doesn't make sense - it's literally x = new_value (and scrap anything that was in x), if the new_value is a modified copy of x then for many cases this would be OK.

In the case of appending to the list, I would expect that to be done via a function called add_ingredient or similar.

Sophiak113 replied:
ok, i will change set ingredients to set x=newvalue and add an add ingredient function which appends

lachlanin4 commented on recipe.py:
```diff
@@ -34,9 +34,13 @@ def get_ingredients(self):
+    
     def remove_ingredient(self,ingredient):
         self._ingredients.remove(ingredient)
+        #Do I need to prevent this removing duplicates if I have already prevented duplication with the set ingredients validation
```
I'm not sure about removing a single ingredient over creating a new modified list of ingredients and simply setting rather than using remove - hmm I will need to have a think.

Sophiak113 replied:
So it is possible that remove_ingredient() will not be needed and we will ask the user to create a new list and set that? Should I remove remove_ingredient()? Would a new function to set ingredients list need to be created?

lachlanin4 replied:
It depends on how the user interface works to a degree, since it's principally at that point one of two things could happen, the UI could create it's own temporary list of changes/additions by the user and then use the temporary list to create or modify a Recipe class instance or it could simply use functionality within a pre generated instance of a Recipe class to do the work. Some might think that the UI should decide when values are correct and should be verified (through application of a save or apply button for example - if you ever had a spell checker that goes berserk before you have even finished the word you can probably see what I mean) others would suggest that you need to check as you are going on as to whether values being entered are correct and as such you need to create the class instance and apply the changes continuously to see if they are valid - but equally if you are making continuous changes then this carries its own problems like the spelling correction example. It's worthwhile keeping it in, things should become clearer as the GUI develops.

Sophiak113 replied:
ok i will keep in remove ingredient

lachlanin4 commented on recipe.py:
```diff
@@ -35,16 +35,19 @@ def get_ingredients(self):
-        print(self.get_ingredients()) #return a pretty string instead
-
+        for ingredient in self._ingredients:
+            print(ingredient) #When i use return it only returns the first value tomato I am not sure why
```
With the new ingredient class this will change,

Sophiak113 replied:
Okay so display_ingredients() will still exist but code underneath it will change or is display ingredients not needed?

lachlanin4 replied:
Yep, ultimately it's the UI that will display to the user, but pretty printing should be added to classes to help with debug and test. There is a way of baking in "sensible" printing of class contents by default (so you can simplify to print(myClassInstance) rather than print(myClassInstance.fancy_print()) ) but it doesn't make sense in all cases and can be misleading at this level.

Sophiak113 replied:
ok, I will not use it in for this class

Sophiak113 committed:
removed functions, added function, hashed out code

removed display and pretty format function, changed set ingredients and took out validation, added add ingredients

lachlanin4 commented on recipe.py:
```diff
@@ -32,21 +32,19 @@ def set_calories_per_portion(self, calories_per_portion:int):
-        if len(set(self._ingredients)) != len(self._ingredients):
-            self.remove_ingredient(ingredients) #I have tested this and it only removes the recently added ingredient
-            return False
+        # if len(set(self._ingredients)) != len(self._ingredients):
```
extend should be used if ingredients are presented as a list, it would best that the type be added in the call to make the point clear.

lachlanin4 commented on recipe.py:
```diff
@@ -32,21 +32,19 @@ def set_calories_per_portion(self, calories_per_portion:int):
     def remove_ingredient(self,ingredient):
         self._ingredients.remove(ingredient)
         #[x for x in self._ingredients if x is not None] #tried to use list comprehension to get rid of the none value in self._ingredients did not succeed
         #Do I need to prevent this removing duplicates if I have already prevented duplication with the set ingredients validation
```
Arguably the ingredients should be unique, but at this point a degree of permissiveness is required to progress things.

lachlanin4 reviewed (changes requested):
I'm going to attempt to try to consolidate both the recipe and recipemanager in one branch tomorrow to show how I think it should all work together - so hopefully we won't be tied up with further reviews across two pull requests.
